pennylane/labs/dla/cartan_subalgebra.py

```diff
@@ -0,0 +1,269 @@
+    >>> g = [op.pauli_rep for op in g] # optional; turn to PauliSentence for convenience
+    >>> k, m = cartan_decomp(g, even_odd_involution)
+    >>> g = k + m # re-order g to separate k and m
+    >>> adj = qml.structure_constants(g)
+    >>> newg, k, mtilde, h, new_adj = cartan_subalgebra(g, k, m, adj)
+    >>> newg == k + mtilde + h
+    True
+    >>> h
+    [(1+0j) * Z(0) @ Z(1), (-1+0j) * Y(0) @ Y(1), (1+0j) * X(0) @ X(1)]
+
+    We can confirm that these all commute with each other, as the CSA is Abelian (= all operators commute).
+
+    >>> from pennylane.labs.dla import check_all_commuting
+    >>> check_all_commuting(h)
+    True
+
+    We can opt-in to return what we call adjoint vectors of dimension :math:`|\mathfrak{g}|`, where each component corresponds to an entry in (the ordered) ``g``.
+    The adjoint vectors for the Cartan subalgebra are in ``np_h``.
+
+    >>> np_newg, np_k, np_mtilde, np_h, new_adj = cartan_subalgebra(g, k, m, adj, return_adjvec=True)
+
+    We can reconstruct an operator by computing :math:`\hat{O}_v = \sum_i v_i g_i` for an adjoint vector :math:`v` and :math:`g_i \in \mathfrak{g}`.
+
+    >>> v = np_h[0]
+    >>> op = sum(v_i * g_i for v_i, g_i in zip(v, g))
+    >>> op.simplify()
+    >>> op
+    Z(0) @ Z(1)
+
+    For convenience, we provide a helper function :func:`~adjvec_to_op` for the collections of adjoint vectors in the returns.
+
+    >>> from pennylane.labs.dla import adjvec_to_op
+    >>> h = adjvec_to_op(np_h, g)
+    >>> h
+    [Z(0) @ Z(1), (-1+0j) * (Y(0) @ X(1)), (-1+0j) * (X(0) @ Y(1))]
+
+    .. details::
+        :title: Usage Details
+
+        Let us walk through an example of computing the Cartan subalgebra. The basis for computing
+        the Cartan subalgebra is having the Lie algebra :math:`\mathfrak{g}`, a Cartan decomposition
+        :math:`\mathfrak{g} = \mathfrak{k} \oplus \mathfrak{m}` and its adjoint representation.
+
+        We start by computing these ingredients using :func:`~cartan_decomp` and :func:`~structure_constants`.
+        As an example, we take the Lie algebra of the Heisenberg model with generators :math:`\{X_i X_{i+1}, Y_i Y_{i+1}, Z_i Z_{i+1}\}`.
+
+        >>> from pennylane.labs.dla import lie_closure_dense, cartan_decomp
+        >>> n = 3
+        >>> gens = [X(i) @ X(i+1) for i in range(n-1)]
+        >>> gens += [Y(i) @ Y(i+1) for i in range(n-1)]
+        >>> gens += [Z(i) @ Z(i+1) for i in range(n-1)]
+        >>> g = lie_closure_dense(gens)
+
+        Taking the Heisenberg Lie algebra, we can perform the Cartan decomposition. We take the :func:`~even_odd_involution` as a valid Cartan involution.
+        The resulting vertical and horizontal subspaces :math:`\mathfrak{k}` and :math:`\mathfrak{m}` need to fulfill the commutation relations
+        :math:`[\mathfrak{k}, \mathfrak{k}] \subeq \mathfrak{k}`, :math:`[\mathfrak{k}, \mathfrak{m}] \subeq \mathfrak{m}` and :math:`[\mathfrak{m}, \mathfrak{m}] \subeq \mathfrak{k}`,
```

Comment:
Here the rendered version seems to have some issues

Reply:
✅